rustls/src/crypto/mod.rs

```diff
@@ -304,16 +314,67 @@ pub trait ActiveKeyExchange: Send + Sync {
     /// mis-encoded, or an invalid public key (such as, but not limited to, being
     /// in a small order subgroup).
     ///
+    /// If the key exchange algorithm is FFDHE, the result must be left-padded with zeros,
+    /// as required by [RFC 8446](https://www.rfc-editor.org/rfc/rfc8446#section-7.4.1)
+    /// (see [`complete_for_tls_version()`](Self::complete_for_tls_version) for more details).
+    ///
     /// The shared secret is returned as a [`SharedSecret`] which can be constructed
     /// from a `&[u8]`.
     ///
     /// This consumes and so terminates the [`ActiveKeyExchange`].
     fn complete(self: Box<Self>, peer_pub_key: &[u8]) -> Result<SharedSecret, Error>;
 
+    /// Completes the key exchange for the given TLS version, given the peer's public key.
+    ///
+    /// Note that finite-field Diffie–Hellman key exchange has different requirements for the derived
+    /// shared secret in TLS 1.2 and TLS 1.3 (ECDHE key exchange is the same in TLS 1.2 and TLS 1.3):
+    ///
+    /// In TLS 1.2, the calculated secret is required to be stripped of leading zeros
+    /// [(RFC 5246)](https://www.rfc-editor.org/rfc/rfc5246#section-8.1.2).
+    ///
+    /// In TLS 1.3, the calculated secret is required to be padded with leading zeros to be the same
+    /// byte-length as the group modulus [(RFC 8446)](https://www.rfc-editor.org/rfc/rfc8446#section-7.4.1).
+    ///
+    /// The default implementation of this method delegates to [`complete()`](Self::complete) assuming it is
+    /// implemented for TLS 1.3 (i.e., for FFDHE KX, removes padding as needed). Implementers of this trait
+    /// are encouraged to just implement [`complete()`](Self::complete) assuming TLS 1.3, and let the default
+    /// implementation of this method handle TLS 1.2-specific requirements.
+    ///
+    /// This method must return an error if `peer_pub_key` is invalid: either
+    /// mis-encoded, or an invalid public key (such as, but not limited to, being
+    /// in a small order subgroup).
+    ///
+    /// The shared secret is returned as a [`SharedSecret`] which can be constructed
+    /// from a `&[u8]`.
+    ///
+    /// This consumes and so terminates the [`ActiveKeyExchange`].
+    fn complete_for_tls_version(
+        self: Box<Self>,
+        peer_pub_key: &[u8],
+        tls_version: &SupportedProtocolVersion,
+    ) -> Result<SharedSecret, Error> {
+        if tls_version.version != ProtocolVersion::TLSv1_2 {
+            return self.complete(peer_pub_key);
+        }
+
+        let group = self.group();
+        let mut complete_res = self.complete(peer_pub_key)?;
+        let kx = group
+            .key_exchange_algorithm()
+            .expect("`ActiveKeyExchange` with invalid group");
```

Comment:
Add a comment on why this is statically impossible, or make it yield an `Error` instead?

Reply:
It's not statically impossible. It can happen if a `CryptoProvider` has incorrectly behaving `ActiveKeyExchange`s in it. I made it panic because an `ActiveKeyExchange` should never behave like this.
Should I keep this behavior, or should I make it return an `Error` instead?

Reply:
Per @ctz's comment, I made `key_exchange_algorithm()` total, so `expect` is not needed here anymore.